### DISCUSSION TITLE
POSRES_FC update

### DISCUSSION THREAD

**Discussion by Vini161** [marrink-lab/vermouth-martinize#639] [Q&A]
I want to slowly release FC in my simulation, the -pf flag gives itp with the value i choose but it doesn't have ifdef POSRES, is it available somewhere else or should I make itp with different FC everytime?

**Comment by pckroon** [comment]
Hmmn, it seems that currently that's indeed the only way.
@csbrasnett How do you feel about changing the `-pf` default to (e.g.) `POSRES_FC`, and adding a default value (e.g. 1000) to the top of the itp file? Is it then possible to override the itp value with a value from the mdp?

**Reply by Vini161** [reply]
I tried assigning a constant such as POSRES_FC, but seems to only accept float in -pf, otherwise using POSRES_FC in mdp without it being defined in molecule_0.itp gives POSRES_FC macro not used error in gmx

**Reply by pckroon** [reply]
That's what I also tried to do to get a workaround for you ;)
Maybe set `-pf 1234567890` and then use something like `sed` to swap that value?

**Reply by Vini161** [reply]
Ahhh! I will try that! thanks!

**Comment by csbrasnett** [comment]
So, I like the suggestion of writing some master value. 

My advancement to allow more flexibility is to try to allow something along the lines of the [ts2cg tutorial](https://github.com/marrink-lab/TS2CG/tree/master/tutorial/mixedbilayer). I don't think it's possible to overwrite an itp from the mdp. But, if people were fine with it and understood what was happening, you could allow for a series of `#ifdef ...` to be written, eg. `#ifdef posres1000`, `#ifdef posres500`, whatever, and control from the mdp that way. We'd have to allow for `-pf` to be given multiple times, but I don't think it should be too hard to do? 

Otherwise I think the two options to control this are either 1) `sed` as you suggest with some identifiable string or 2) use a custom python script to load the itp with vermouth, remove the `[ position_restraints ]` directive (assuming there aren't any other position restraints in the system), and add new interactions with whatever force constant you want. But if you've not written scripts with vermouth like that it's probably a bit of a pain to do!

**Reply by csbrasnett** [reply]
In fact, I'd guess that if you defined the master `POSRES_FC` at the top of the itp somehow then you could do the `#ifdef`s there too. But that might break martinize more than just writing multiple `#ifdef`s within a single directive

**Reply by pckroon** [reply]
How about an `#ifndef POSRES_FC; POSRES_FC=1000; #endifndef` or whatever the syntax is? I seem to vaguely recall seeing something like that somewhere at some point. Perfectly accurate, detailed and precise information here ;)

**Reply by csbrasnett** [reply]
Per some playing around and the gromacs forum post below, the lines we'd need would be:

```
#ifndef POSRES_FC
#define POSRES_FC 1000
#endif
```
which can be written anywhere in the itp file

I guess the question is what's best for users generally? I could imagine this might be confusing without further explanation, but then again, we expect people to know to specify `define = -DFLEXIBLE` in their mdp file for minimization already.

Maybe if it's changed to write this, along with a logger.INFO that'd be best. But, I can't actually work out atm how the lines could/should actually be easily written into an itp

**Reply by pckroon** [reply]
Yes, this is what I had in mind. And I'm perfectly fine with assuming users know how gromacs works, in this specific case the `define` option in MDPs. 
It would not be a bad idea to document which variables are used in martinize2-made itps files, but not super high prio to me.

In `vermouth/processors/apply_posres.py:apply_posres`, I would suggest adding something sensible to `molecule.meta` (for example (`molecule.meta['define'] = {'POSRES_FC': 1000}`), but the itp writer currently doesn't pick this up. So we'd need to add some logic to `vermouth/gmx/itp.py:140`, right before `[ moleculetype ]` gets written.

```py
for name, val in molecule.meta.get('define', {}).items():
    outfile.write(f'#ifndef {name}')
    outfile.write(f'#define {name} {value}')
    outfile.write(f'#endif')
```
Or something along those lines.

**Comment by csbrasnett** [comment]
Just to make the link, I've been playing around with how implementing this might be possible and have started an associated discussion in the gromacs forum [here](https://gromacs.bioexcel.eu/t/defining-position-restraints-in-itp-mdp/11030)

**Comment by csbrasnett** [comment]
Hi @Vini161 just fyi, this feature is now available in the main branch if you install it!

**Comment by Vini161** [comment]
Thanks
